cluster/service/cluster_service_blackbox_test.go

```diff
@@ -50,6 +53,50 @@ func (s *ClusterServiceTestSuite) TestCreateOrSaveClusterOK() {
 	verifyClusters(s.T(), append(osoClusters, osdClusters...), s.Configuration.GetClusters())
 }
 
+func (s *ClusterServiceTestSuite) TestCreateOrSaveCluster() {
+
+	s.T().Run("ok", func(t *testing.T) {
+		// given
+		clustr := newTestCluster()
+		// when
+		err := s.Application.ClusterService().CreateOrSaveCluster(context.Background(), &clustr)
+		// then
+		require.NoError(t, err)
+		assert.NotNil(t, clustr.ClusterID)
+	})
+
+	s.T().Run("failure", func(t *testing.T) {
+		// given
+		clustr := newTestCluster()
+		clustr.Name = ""
+		// when
+		err := s.Application.ClusterService().CreateOrSaveCluster(context.Background(), &clustr)
+		// then
+		require.Error(t, err)
+
+	})
```

Comment:
this is going to be a duplicate of the validation tests that I did in https://github.com/fabric8-services/fabric8-cluster/pull/41/files#diff-e9581e303b0b0335c4bed88053faff86R20. Here, I just wanted to make sure that in case of a validation error, the service returns an error

Reply:
You don't know if the validation is actually called. I would rather test the whole service which will test the internal validation function indirectly then relaying of testing implementation details of the service.

Reply:
Your validate() function may work perfectly and all tests will pass but if this function is not used properly in the service or not used at all then the service won't work as expected but your tests won't catch it.
We have to make sure we test all exported functions very well with all possible outputs and usecases. Testing internal functions and implementation details is optional. Sometimes we have to test them because it's very tricky to test the exposed functions fully, so, it's better to test at least some the implementation parts (== not exposed functions) instead of the API (== exposed functions) than nothing. But it's not the case.

Reply:
fair enough, I'll refactor the tests, because there's no point in having duplicate tests here, then

Reply:
addressed in e54ba34